hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionImpl.java

```diff
@@ -222,70 +205,76 @@ public void prepare() throws TransactionException {
+     * each logrecord needs to prepare its content (e.g. by acquiring locks). This takes time.
+     *
+     * If a transaction is local or if there is 1 or 0 items in the transaction log, instead of
+     * preparing, we are just going to try to commit. If the lock is still acquired, the write
+     * succeeds, and if the lock isn't acquired, the write fails; the same effect as a prepare
+     * would have.
+     *
+     * @return true if {@link #prepare()} is required.
+     */
+    public boolean requiresPrepare() {
+        if (transactionType == LOCAL) {
+            return false;
+        }
 
-            Future f = operationService.invokeOnTarget(TransactionManagerServiceImpl.SERVICE_NAME,
-                    new ReplicateTxOperation(transactionLog.getRecordList(), txOwnerUuid, txnId, timeoutMillis, startTime),
-                    backupAddress);
-            futures.add(f);
+        if (transactionLog.size() <= 1) {
+            return false;
         }
-        waitWithDeadline(futures, timeoutMillis, MILLISECONDS, RETHROW_TRANSACTION_EXCEPTION);
-        // todo: why clear? The futures will be gc'd
-        futures.clear();
+
+        return true;
     }
 
     @Override
     public void commit() throws TransactionException, IllegalStateException {
         try {
-            if (transactionType.equals(TWO_PHASE) && state != PREPARED) {
-                throw new IllegalStateException("Transaction is not prepared");
-            }
-            if (transactionType.equals(TransactionType.LOCAL) && state != ACTIVE) {
+            if (transactionType == TWO_PHASE) {
+                if (transactionLog.size() > 1) {
+                    if (state != PREPARED) {
+                        throw new IllegalStateException("Transaction is not prepared");
+                    }
+                } else {
+                    // when a transaction log contains less than 2 items, it can be committed without preparing
+                    if (state != PREPARED && state != ACTIVE) {
+                        throw new IllegalStateException("Transaction is not prepared or active");
+                    }
+                }
+            } else if (transactionType == LOCAL && state != ACTIVE) {
                 throw new IllegalStateException("Transaction is not active");
             }
+
             checkThread();
             checkTimeout();
             try {
                 state = COMMITTING;
-
                 List<Future> futures = transactionLog.commit(nodeEngine);
-
-                // We should rethrow exception if transaction is not TWO_PHASE
-                ExceptionHandler exceptionHandler = transactionType.equals(TWO_PHASE)
-                        ? commitExceptionHandler : FutureUtil.RETHROW_TRANSACTION_EXCEPTION;
-                waitWithDeadline(futures, COMMIT_TIMEOUT_MINUTES, MINUTES, exceptionHandler);
-
+                waitWithDeadline(futures, COMMIT_TIMEOUT_MINUTES, MINUTES, RETHROW_TRANSACTION_EXCEPTION);
```

Comment:
a question:
we were not throwing any exception while committing but we were logging.
with this we are changing that, how will it affect a user ?
```
try {
    do something
    commit
} catch(Exception) {
    rollback
}
```
What will happen for the above case ? user will get the exception and will try to rollback it, will it throw exception too ?